src/tx/builder/schema.ts

```diff
@@ -403,15 +509,47 @@ const NAME_TRANSFER_TX = [
   TX_FIELD('ttl', FIELD_TYPES.int)
 ]
 
+type BuildTxArgBySchemaType<SchemaType> = SchemaType extends 'int'
+  ? string | number | BigNumber
+  : SchemaType extends 'id'
+    ? string
+    : SchemaType extends typeof Field
+      ? Parameters<SchemaType['serialize']>[0]
+      : never
+
+// based on https://stackoverflow.com/a/50375286/6176994
+type UnionToIntersection<Union> =
+  (Union extends any ? (k: Union) => void : never) extends ((k: infer Intersection) => void)
+    ? Intersection : never
+
+type BuildTxArgBySchema<SchemaLine> =
+  UnionToIntersection<
+    SchemaLine extends ReadonlyArray<infer Elem>
+      ? Elem extends readonly [string, string | Field] ? { [k in Elem[0]]: BuildTxArgBySchemaType<Elem[1]> } : never
+      : never
+    >
+
 const NAME_REVOKE_TX = [
-  ...BASE_TX,
-  TX_FIELD('accountId', FIELD_TYPES.id, 'ak'),
-  TX_FIELD('nonce', FIELD_TYPES.int),
-  TX_FIELD('nameId', NameId),
-  TX_FIELD('fee', FIELD_TYPES.int),
-  TX_FIELD('ttl', FIELD_TYPES.int)
-]
+  // ...BASE_TX,
+  ['tag', FIELD_TYPES.int],
+  ['VSN', FIELD_TYPES.int],
+  ['accountId', FIELD_TYPES.id, 'ak'],
+  ['nonce', FIELD_TYPES.int],
+  ['nameId', NameId],
+  ['fee', FIELD_TYPES.int],
+  ['ttl', FIELD_TYPES.int]
+] as const
+export interface TxNameRevoke extends BuildTxArgBySchema<typeof NAME_REVOKE_TX> {}
```

Comment:
This example is for buildTx, the similar thing needs to be done for unpackTx, it can't be the same because, for example, as `amount` we accepting `string | number | BugNumber`, but returning the only `string`.
Also the code is broken now, because other parts should be rewrited as well.